pkg/ctrlkit/shared.go

```diff
@@ -57,12 +58,22 @@ func (s *sharedAction) Run(ctx context.Context) (ctrl.Result, error) {
 	// Start a new goroutine to do this.
 	go s.once.Do(func() {
 		defer close(s.done)
+		defer func() {
+			if r := recover(); r != nil {
+				s.panics <- r
+			}
+		}()
 		s.result, s.err = s.inner.Run(ctx)
 		s.done <- true
 	})
 
 	// Wait on the done channel. Memory barrier should also be carried here.
-	<-s.done
+	select {
+	case msg := <-s.panics:
+		panic(msg)
+	case <-s.done:
+
+	}
```

Comment:
The `Run` method of the `sharedAction ` can be invoked multiple times and always get the same result. However, the `<-s.panics` can only receive value once and will not get the panic value in the following invocations. I think we need to fix this.